### ISSUE TITLE
Passing context to function validation decorators and accessing value

### ISSUE THREAD

**Issue by jbmusso** [fcracker79/pycomb#10]
Greetings, and thanks for porting tcomb to Python 👍.

I'm currently validating function input and output using decorators. Here's a simple example with an `add()` function expected to work on natural numbers:

```python
def is_positive(x: int) -> bool:
    return x >= 0

def is_integer(x: int) -> bool:
    return x % 1 == 0

def is_natural(x: int) -> bool:
    return is_positive(x) and is_integer(x)

natural_numbers = combinators.subtype(
    combinators.Int, is_natural, example=42, name="Natural number"
)


@combinators.function(natural_numbers, natural_numbers)
@returning(natural_numbers)
def add(a: int, b: int) -> int:
    return a + b
```

I dug through the documentation and code and if I understand correctly, a custom context cannot be provided to a `@decorator` when performing validation. Also, it appears that only the type can be accessed in the context, and not the value.

Goal: return a more detailed error message indicating which specific value failed the validation, along with its type, all using a custom message.
The most verbose message I can get at the moment is

`Error on Natural number: expected Natural number but was int`.

I'm trying to get something like:

`Error on Natural number: expected Natural number but was int (value: -1)`

Would you recommend anything or does this require additional work? If so, I could be working on a pull request. Thanks for your time.

**Comment by fcracker79** [comment]
Greetings, and thank you for using PyComb!

You are totally correct: it is not possible to pass a custom context to decorators (sorry, my mistake, I completely forgot to implement it), and observers do not receive objects but just types.

Concerning the former, it would be great if you could do a PR. Or I can do it myself, it would not take that much.

Concerning the latter, the idea was not to delegate the string message to the final user, as it could have been risky (buggy `__str__` method, very verbose string output, unexpressive type indication...). 
On the other hand, though, I understand your need and I am thinking of how to implement it without introducing a breaking change: as you might have seen in the code, all the combinators pass only a `type(x)` in `found_type` keyword argument, which would not be such an issue if there were not custom combinators out there. In order to preserve the behaviour of custom combinators, it would be necessary to introduce an additional keyword argument such as `found_value` in observers, but avoid to pass it in observers that do not support it.

Alternatively, we might think of setting the current object in the context: this could be easier as the property can be added to the `ValidationContext` superclass, would not be a breaking change and would impact only to observers that are specifically written to fetch the current validating object, the other ones would simply ignore that value. Finally, custom combinators would not push the validating data in the context, but this is not an issue, as only users interested in the validating value would be forced to do that.

Thanks again for your feedback.
Please let me know what you think.

**Comment by fcracker79** [comment]
Meanwhile, you might have a look at 0e5bd5cbd8a8f03c1be4bf4f925da4a0ac1cd988.
It would be nice if you could provide some feedback.

**Comment by fcracker79** [comment]
Released 0.1.17 version with the above implementation.
Hope it fits your needs.

**Comment by jbmusso** [comment]
Hello, sorry for the late reply - last few days have been very busy. Thanks for pushing this new release, I'll check this over the weekend and let you know how it goes!
